### PR TITLE
[GEOT-6579]: added array filter function, that can be used to create …

### PR DESCRIPTION
**Pull request by mbarto**
…array literals in ECQL filters

<Include a few sentences describing the overall goals for this Pull Request>

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geotools/geotools/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.
- [ ] The changes are not breaking the build in downstream projects using SNAPSHOT dependencies, GeoWebCache and GeoServer.

The following are required only for core and extension modules (they are welcomed, but not required, for unsupported modules):
- [x] There is an issue in [Jira](https://osgeo-org.atlassian.net/projects/GEOT) describing the bug/task/new feature (a notable exemptions is, changes not visible to end users). The ticket is for the GeoTools project, if the issue was found elsewhere it's a good practice to link to the origin ticket/issue.
- [x] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[GEOT-XYZW] Title of the Jira ticket"
- [x] New unit tests have been added covering the changes
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] This PR passes the [QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html) (QA checks results will be reported by travis-ci after opening this PR)
- [ ] Documentation has been updated accordingly.

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.

### CODE REVIEW

File: modules/library/main/src/main/java/org/geotools/filter/function/ArrayFunction.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ */
+package org.geotools.filter.function;
+
+import static org.geotools.filter.capability.FunctionNameImpl.parameter;
+
+import org.geotools.filter.FunctionExpressionImpl;
+import org.geotools.filter.capability.FunctionNameImpl;
+import org.opengis.filter.capability.FunctionName;
+
+public class ArrayFunction extends FunctionExpressionImpl {
```
Comment (jodygarnett) [inline modules/library/main/src/main/java/org/geotools/filter/function/ArrayFunction.java:25]: public class javadoc

File: modules/library/main/src/main/java/org/geotools/data/util/ArrayConverterFactory.java [hunk 2]
```diff
@@ -55,4 +58,18 @@ public Converter createConverter(Class<?> source, Class<?> target, Hints hints)
...
+
+    private static class ArrayToArrayConverter implements Converter {
+        @Override
+        public <T> T convert(Object source, Class<T> target) throws Exception {
+            int arrLength = Array.getLength(source);
+            Class<?> elementType = target.getComponentType();
+            Object result = Array.newInstance(elementType, arrLength);
+
+            for (int count = 0; count < arrLength; count++) {
+                Array.set(result, count, Converters.convert(Array.get(source, count), elementType));
```
Comment (jodygarnett) [inline modules/library/main/src/main/java/org/geotools/data/util/ArrayConverterFactory.java:70]: This looks great, is it worth checking for `null` here (returned when converters cannot make the conversion) or do you want to just let it ride?

Comment (mbarto) [reply modules/library/main/src/main/java/org/geotools/data/util/ArrayConverterFactory.java:70]: @jodygarnett good catch. I think we could check the source element, if it's null then set the target array element to null, if the source is not null and the converted is (error in conversion) we can return null for the whole target array (to notify a conversion error). What do you think?

Comment (jodygarnett) [reply modules/library/main/src/main/java/org/geotools/data/util/ArrayConverterFactory.java:70]: Yes that is the correct approach.

Note: We did experiment with returning exceptions for this kind of failure, but the resulting exceptions would end up user visible (resulting in confused unhappy complaints). We ended up with this *return null* approach which IMHO ends up being too silent on details ... a tradeoff.

**Pull request merged**